### DISCUSSION TITLE
diff flow without index

### DISCUSSION THREAD

**Discussion by ghost** [jj-vcs/jj#2068] [Q&A]
I see that JJ has no index:

https://github.com/martinvonz/jj/blob/main/docs/git-comparison.md

I am OK with that, but I am wondering how to adapt my current flow. usually with Git, a commit will involve multiple files. so when preparing a commit, I will:

1. `git diff`
2. once I confirm a specific file is OK, then `git add file`
3. loop back to step 1 until all files are done, OR maybe just commit some of the files only

the benefit of this is that as I make progress, the diff gets smaller. does JJ have some similar workflow to this?

**Comment by ilyagr** [comment]
Could I ask you to be a guinea pig for our new FAQ? See https://github.com/martinvonz/jj/blob/main/docs/FAQ.md#can-i-prevent-jujutsu-from-recording-my-unfinished-work-im-not-ready-to-commit-it.

Let me know if it helps, and feel very welcome to ask follow-up questions, especially if it doesn't address your question (but also in any case).

**Reply by ghost** [reply]
I am really having trouble understanding the language used on the documentation here. for example:

> However, you can easily record intermediate drafts of your work.

whats the difference between an intermediate draft and just a draft? aren't all drafts intermediate by definition?

> If you think you might want to go back to the current state of the working-copy commit, simply use `jj new`.

the current state of the working copy commit is just the current state of all the files. so it seems that `jj new` would never do anything.

> Then future edits will go into a new working-copy commit on top of the now former working-copy commit.

I have no idea what this means. it seems to suggest to make two commits, whereas in Git I would have made no commits so far, and only so far added items to the index.

**Reply by arxanas** [reply]
In Git, you should instead consider the equivalent workflow of creating a new (empty) commit and then repeatedly `git commit --amend`ing into it (instead of using `git add`) and using `git show` (instead of `git diff`). Then, you can directly transfer the same workflow into jj by creating a new commit with `jj new` and amending into it using one of the commands listed in that document.

**Reply by ghost** [reply]
`git show` is going to show whats in the commit - I want to show what HAS NOT been committed, which is what `git diff` does

**Reply by arxanas** [reply]
My mistake. I mixed up two alternative workflows; we wouldn't use `git show` in that case. For the jj workflow, we first need to create an extra parent commit to amend into. I usually do this with `jj split`. After we have the extra parent commit, `jj diff` would behave as we wanted and show the changes in the working copy (commit) compared to the parent commit, i.e. the "uncommitted" changes. (`jj show` would also work the same in this case, hence my mistake.) We would use one of the other commands to successively move changes from the working copy into the parent commit.

**Reply by ghost** [reply]
OK, but whats the equivalent of `git add file`?

**Reply by arxanas** [reply]
The equivalent would be `jj squash` (also has `jj amend` as an alias if you prefer that like me). This would let you move changes from the working copy commit into the parent commit.

**Reply by ghost** [reply]
hm OK, I guess thats just too different for my liking. the benefit of the index for me, is its a clear delineation between whats "done" and whats not done. whatever is a commit is "done", and anything in the index is "partially done" and changes in the working directory are "not done".

with this, it seems any file change is already "committed", so it seems you have no way to understand the different states of "done-ness". or maybe I just dont understand it.

**Reply by arxanas** [reply]
> with this, it seems any file change is already "committed",

Yes, the working copy is always committed. For me, the phases are typically determined by commit message:

- A real commit message means that it's "done".
- A commit message starting with e.g. `wip:`. means that it's "partially done".
  - Unlike the staging area, I may have multiple successive commits marked as `wip`. I'll promote them to "real' commits later by updating the commit messages (or drop them, or squash their changes into another commit, etc.).
- No commit message means that it's "not done".
  - I only use this for working copy commits (i.e. I never have two successive commits without commit messages).
  - jj will also prevent you from pushing commits without commit messages when you use `jj git push`.

There has been discussion of adding Mercurial-style phases. I suppose you could map the "secret"/"draft"/"public" phases onto these three states.

**Reply by ghost** [reply]
> * A commit message starting with e.g. `wip:`. means that it's "partially done".

thats what I was afraid of. this just seems like re-inventing the index, using an idiom instead of an integrated part of the tool. so it appears that the index was removed, but people are still using it in an ad hoc way. I prefer to just use a tool that supports it properly, so maybe JJ is not right for me.

**Reply by thoughtpolice** [reply]
You'd normally just do `jj amend` frequently, and if you need it, `jj amend -i`, which is basically like `git add -p`, but it moves the selected changes from the current working-copy-commit into the parent commit.

The thing is, at some point when you're writing code in an editor, there is always a point where you come to a stopping point, say "I'm done with this little thing", and you run a command like `git add` in order to stage your changes into the index, whether or not it's interactive, a single file, all touched files, etc. You do the same thing in `jj`; you reach a stopping point and write `jj amend file.ext` for example to amend the changes in `file.ext` into the parent.

The real distinction is what "verbs" operate on what "nouns". In Git, there are lots of nouns: the index, the stash, the `HEAD`, etc. The verbs are things like `commit` or `stash`. Internally these use the same objects in the underlying data model (blobs), but externally they are different objects in the UX. There are many nouns in Git, as well as many verbs. In `jj`, the stash, the index, *and* a commit are all the same underlying object, both in the data model and in the UX. This means there are less nouns (only commits), and there are fewer verbs, because they need to operate on less objects.

So what I normally do is something like this:

- `jj new main` to create a new working copy commit on top of `main`
- Write a little code. 
- At some point I say, "This is basically complete"
- I do `jj commit -m "Really cool thing"`, which updates the commit message for the working copy commit, then creates a new commit on top of it.
  - This is 100% equivalent to running `jj describe` followed by `jj new`.
- At this point, a few things might happen.
  - I might be done. I might `jj git push` this change now.
  - I might find a small flaw I need to incorporate.
    - I make that change
    - I run `jj amend file.ext`
    - This is sort of like running `git add` or `git add -p`, except instead of putting the changes into the index, it just puts them into the parent commit.
    - I will eventually finish, and now go back to the other option: I will probably `jj git push` changes
  - I might decide that some changes should be split out into two commits
    - I could run `jj split` on that commit in order to split it into two commits.

You don't need the index in this model, because all behavior is modeled by operations (verbs) on commits (the most fundamental noun).

Similarly, you don't need things like `stash` in this model. If you want to stash changes and get them out of the way &mdash; just run `jj new <name-of-branch>` or whatever, and you're done. Any changes you wrote will be put into the working copy commit, then a *new* working copy commit will be created on top of whatever you want. You can `jj edit` to go back to any commit at any time.

Ultimately I don't think your workflow is very different. It's a matter of embracing the fact there are less nouns. The operations are in a lot of ways somewhat similar, in my experience. It wasn't too hard to adopt once I got this mindset.

**Reply by arxanas** [reply]
If you require the index, then jj is probably not for you. The intention is to 1) reduce the number of working copy states, such as demonstrated in https://spderosso.github.io/onward13.pdf:

<img width="887" alt="Screenshot 2023-08-13 at 21 05 41" src="https://github.com/martinvonz/jj/assets/454057/ae3d89cb-03a1-4ef5-9801-d8cd18f30232">

and 2) allow you to use the same set of tools to interact with the index as with regular commits. For example, in Git, you can't take a staged change and move it directly into an arbitrary other commit, but you can do so with `jj move`.

There are some other advantages to jj. For example, conflicts can be stored directly in commits, which mean that rebases always succeed and you can choose to resolve the conflicts at a later time. This can be helpful when rebasing a large stack of commits, since you can go back/forward in the stack as desired and update change merge conflict resolutions as you get familiar with the conflicts.

**Reply by ghost** [reply]
> You'd normally just do `jj amend` frequently, and if you need it, `jj amend -i`, which is basically like `git add -p`, but it moves the selected changes from the current working-copy-commit into the parent commit.

I dont understand the terms "working copy commit" or "parent commit" in this context.

> The thing is, at some point when you're writing code in an editor, there is always a point where you come to a stopping point, say "I'm done with this little thing", and you run a command like `git add` in order to stage your changes

I use `git add` when I am partway done with a commit, that way it no longer shows in the diff. I am not sure how to do that with JJ

> You do the same thing in `jj`; you reach a stopping point and write `jj amend file.ext` for example to amend the changes in `file.ext` into the parent.

right but you're committing at that point, so how do I tell the difference between partially done work and really done work? with JJ its all commits, with Git I know that commits are "done", but index is "partially done".

> * `jj new main` to create a new working copy commit on top of `main`

this is confusing, you are creating a commit before doing anything. also what happens if you forget this command and just start editing?

> * I do `jj commit -m "Really cool thing"`, which updates the commit message for the working copy commit, then creates a new commit on top of it.

right, but since this is sort of like the Git index, how can I tell the difference between this possibly unfinished commit, and the normal "done" commits?

> You don't need the index in this model, because all behavior is modeled by operations (verbs) on commits (the most fundamental noun).

see above.

> It's a matter of embracing the fact there are less nouns. The operations are in a lot of ways somewhat similar, in my experience. It wasn't too hard to adopt once I got this mindset.

see above.

**Reply by arxanas** [reply]
> I dont understand the terms "working copy commit" or "parent commit" in this context.

The "working copy commit" is just the commit that is currently checked out. However, in jj, that commit is automatically amended to contain the working copy contents whenever you run a `jj` command. If we were converting your workflow to jj, we would consider it to be the equivalent of unstaged changes.

The significance is that we only have commands that move changes from one commit to another, and no commands that move unstaged changes into the staging area (since there are technically no unstaged changes or staging area).

> I use git add when I am partway done with a commit, that way it no longer shows in the diff. I am not sure how to do that with JJ

In this example workflow where we simulate the staging area, we create an extra commit with `jj new` before we start working (representing the staging area) and then use `jj squash` (also has synonym `jj amend`) to move a file from the working copy commit ("unstaged changes") to the parent commit ("staging area"). Then `jj diff` should do the right thing because it'll show the contents of the working copy commit.

> right but you're committing at that point, so how do I tell the difference between partially done work and really done work? with JJ its all commits, with Git I know that commits are "done", but index is "partially done".

You have to find a way that works for you 🤷. Most people use commit messages. The working copy commit has no commit message by default, so the case of logically-unstaged changes is easy. You could create an alias that quickly creates a commit with the message `staging` or something like that for your own benefit. You could also use the presence of a branch to infer whether a commit is done or not, depending on your workflow. 

We could also ask the rhetorical question: in Git, how would you create multiple "partially done" commits that exist simultaneously? The index can't represent this, so you have to use commits anyways when you get to this point. (I do this quite frequently in both Git and jj).

You might be able update your default `jj log` template to mark commits with certain messages in different colors or something like that, if you prefer to see "partially-done" commits highlighted visually. I haven't tried it.

> this is confusing, you are creating a commit before doing anything. also what happens if you forget this command and just start editing?

Yes: to best mimic the "staging area" workflow, we are creating an extra commit before doing anything that represents the staging area.

If you forget the command, then it's roughly equivalent to the situation in Git when you start working without staging your previous changes. In this situation in jj, I usually use `jj split` to create a new parent commit and move some of my changes into it.

> right, but since this is sort of like the Git index, how can I tell the difference between this possibly unfinished commit, and the normal "done" commits?

See above re commit messages.

**Reply by ghost** [reply]
thanks for all the responses. hopefully they will help future readers as well. I guess this tool is just not for me. maybe I will understand it better one day. I know people complain about the index, but it seems quite useful to me. its a simple and elegant way to say "here is something that is partially done" while a commit is a way to say "this is done", and the working directory changes are "not done". JJ seems to say:

> everything is a commit and everything is done, but if you're not actually done you need to go back and fix what was not done, and the tooling will not provide an integrated way to delineate "done" from "not done" work, so you'll need to just come up with a best practice.

**Reply by arxanas** [reply]
> everything is a commit and everything is done, but if you're not actually done you need to go back and fix what was not done, and the tooling will not provide an integrated way to delineate "done" from "not done" work, so you'll need to just come up with a best practice.

That sounds about right 😇. I will let you know if jj adopts Mercurial-style phases, because they might solve your problem.

It wasn't mentioned in this thread, but it's probably worth providing this context: I think most jj users use a heavily patch-based workflow, where we will submit multiple commits for review and, based on code review feedback, go back and amend them until they pass review. In this sense, the commits we're working on are never necessarily "done". Maybe that's why we don't have an emphasis on "done"/"partially-done"/"not done" phases — because no local work is ever "done". Usually it's only "done" when it gets merged into an upstream branch (at which point we won't amend/rewrite it because it's public history).

**Reply by martinvonz** [reply]
> > * A commit message starting with e.g. `wip:`. means that it's "partially done".
> 
> thats what I was afraid of. this just seems like re-inventing the index, using an idiom instead of an integrated part of the tool. so it appears that the index was removed, but people are still using it in an ad hoc way. I prefer to just use a tool that supports it properly, so maybe JJ is not right for me.

At least the way I work, there are often commits that I'm working on that are not yet ready to send for review yet. I often give these a message, but I suspect I'm not as consistent as @arxanas about prefixing them with "wip:". But the point is that I would have these commits when using Git too. Don't you? If you don't, then what do you do when you need to switch tasks for a while? I usually have tens of commits of unfinished work.

**Reply by martinvonz** [reply]
> I guess this tool is just not for me. maybe I will understand it better one day.

It's definitely *meant to* be useful for your workflow. I think part of it is about getting over the fact that the working copy gets automatically committed and not thinking that that means that it has to be "done". And where you currently build up the index by incrementally adding file with `git add`, you'd instead do that by incrementally growing a commit. After making some changes in the working copy and you feel ready to build up a "real/done" commit, you'd run `jj split <file>` (or just `jj split` to interactively select changes) to move some changes into a new commit. Then you can `jj squash/amend ([-i] | <files>)` to add more changes into that commit, just like you would have used `git add ([-p] | <files>)`.

If we wanted an index like git's index, we could do that by treating the working-copy commit's parent differently. We could highlight it in the log output, for example (I think `gitk` shows the index as a green dot and the working copy as a red dot, or at least it did that ~12 years ago). It wouldn't be hard to implement that, but I think it's much better to not treat the working-copy commit's parent specially.

> I know people complain about the index, but it seems quite useful to me. its a simple and elegant way to say "here is something that is partially done" while a commit is a way to say "this is done", and the working directory changes are "not done".

> JJ seems to say:
> 
> > everything is a commit

Yes.

> and everything is done

No, that's up to you to keep track of.

>, but if you're not actually done you need to go back and fix what was not done, and the tooling will not provide an integrated way to delineate "done" from "not done" work, so you'll need to just come up with a best practice.

 I personally consider pretty much every commit that's not in a PR unfinished. Even if we had a command to mark a commit as "done", I would probably just send it for review when it was done anyway, so "marked done" and "is part of PR" would probably almost always go hand in hand.

**Reply by joyously** [reply]
From what I can discern, the OP values the way that Git's diff gets smaller as each file is staged, providing a way to keep track of what is recorded and what is not.
Jj would show a smaller diff as each commit is moved out of the working copy commit, wouldn't it?
I would suggest that perhaps it's a good thing to keep fewer changes uncommitted, which is what jj does for you.

**Comment by martinvonz** [comment]
> I see that JJ has no index:
> 
> https://github.com/martinvonz/jj/blob/main/docs/git-comparison.md
> 
> I am OK with that, but I am wondering how to adapt my current flow. usually with Git, a commit will involve multiple files. so when preparing a commit, I will:
> 
> 1. `git diff`
> 2. once I confirm a specific file is OK, then `git add file`
> 3. loop back to step 1 until all files are done, OR maybe just commit some of the files only
> 
> the benefit of this is that as I make progress, the diff gets smaller. does JJ have some similar workflow to this?

Just to make sure, did you see https://github.com/martinvonz/jj/blob/main/docs/git-comparison.md#the-index? Because the second paragraph there was meant to describe exactly your workflow (as far as I understand it anyway). If you missed that section, we should see how we can clarify it.

**Reply by tp-woven** [reply]
Reading that section now, I wonder if it would be useful to also mention that you can `jj squash <file>`, as it's more similar to `git add <file>`? 🤔

**Reply by martinvonz** [reply]
That would make sense to me.
